Extension/src/LanguageServer/colorization.ts

```diff
@@ -0,0 +1,601 @@
+
+    private purgeOldVersionedEdits(): void {
+        let lowerVersion: number = this.lastSemanticVersion;
+        if (lowerVersion > this.lastSyntacticVersion) {
+            lowerVersion = this.lastSyntacticVersion;
+        }
+        let index: number = this.versionedEdits.findIndex((edit) => edit.editVersion > lowerVersion);
+        if (index === -1) {
+            this.versionedEdits = [];
+        } else if (index > 0) {
+            this.versionedEdits = this.versionedEdits.slice(index);
+        }
+    }
+    
+    private updateColorizationRanges(uri: string, syntacticRanges: vscode.Range[][], semanticRanges: vscode.Range[][], inactiveRanges: vscode.Range[]): void {
+        // Dispose of original decorators.
+        // Disposing and recreating is simpler than setting decorators to empty ranges in each editor showing this file
+        this.disposeColorizationDecorations();
+
+        if (inactiveRanges) {
+            this.inactiveRanges = inactiveRanges;
+        }
+        for (let i: number = 0; i < TokenKind.Count; i++) {
+            if (syntacticRanges) {
+                this.syntacticRanges[i] = syntacticRanges[i];
+            }
+            if (semanticRanges) {
+                this.semanticRanges[i] = semanticRanges[i];
+            }
+        }
+
+        let isCpp: boolean = util.isEditorFileCpp(uri);
+        this.createColorizationDecorations(isCpp);
+
+        // Apply the decorations to all *visible* text editors
+        let editors: vscode.TextEditor[] = vscode.window.visibleTextEditors.filter(e => e.document.uri.toString() === uri);
+        for (let e of editors) {
+            this.refreshColorizationRanges(e);
+        }
+    }
+
+    public updateSyntactic(uri: string, syntacticRanges: vscode.Range[][], editVersion: number): void {
+        this.versionedEdits.forEach((edit) => {
+            if (edit.editVersion > editVersion) {
+                for (let i: number = 0; i < TokenKind.Count; i++) {
+                    syntacticRanges[i] = this.fixRanges1(syntacticRanges[i], edit.changes);
+                }
+            }
+        });
+
+        this.updateColorizationRanges(uri, syntacticRanges, null, null);
+        this.lastSyntacticVersion = editVersion;
+        this.purgeOldVersionedEdits();
+    }
+
+    public updateSemantic(uri: string, semanticRanges: vscode.Range[][], inactiveRanges: vscode.Range[], editVersion: number): void {
+        this.versionedEdits.forEach((edit) => {
+            if (edit.editVersion > editVersion) {
+                for (let i: number = 0; i < TokenKind.Count; i++) {
+                    semanticRanges[i] = this.fixRanges1(semanticRanges[i], edit.changes);
+                }
+                inactiveRanges = this.fixRanges1(inactiveRanges, edit.changes);
+            }
+        });
+
+        this.updateColorizationRanges(uri, null, semanticRanges, inactiveRanges);
+        this.lastSemanticVersion = editVersion;
+        this.purgeOldVersionedEdits();
+    }
+}
```

Comment:
I didn't have time to look at all of this file.  I'll need someone else from the team to sign off here.